grakn-graph/src/main/java/ai/grakn/graph/internal/RelationTypeImpl.java

```diff
@@ -66,9 +64,7 @@ public Relation addRelation() {
      */
     @Override
     public Collection<RoleType> hasRoles() {
-        Set<RoleType> roleTypes = new HashSet<>();
-        getOutgoingNeighbours(Schema.EdgeLabel.HAS_ROLE).forEach(role -> roleTypes.add(role.asRoleType()));
-        return roleTypes;
+        return Collections.unmodifiableCollection(cachedHasRoles.get());
```

Comment:
Just so you're clear on this behaviour: even though the resulting collection cannot be modified, it may still change internally. For example:
```
roles = thing.hasRoles();
roles.contains(rt); // False

thing.hasRole(rt);

roles.contains(rt); // True, by magic
```

I don't know if this is _bad_, but it's worth thinking about.